articles/055_ros_parameter_design.md

```diff
@@ -0,0 +1,207 @@
+The full complement of datatypes of different bitdepth and unsigned types are avoided to allow interpretation from text based configuration files.
+
+`bytes` are included to allow the storage of binary blobs.
+It's use is not recommended but can be very convenient, and explicitly supporting it is better than having people try to abuse other datatypes such as strings.
+
+### Required functionality
+
+Each node is responsible for providing the following functionality.
+
+- **Get Parameters**
+  Given a list of parameter names it will return the parameter values.
+- **Set Parameters**
+  Given a list of parameter names, it will request an update of the values subject to validation of the values.
+  The updated values can include unsetting the value.
+  It will provide an API that can atomically update a set of values such that if any of the values fail validation, none of the values are set.
+  The success or failure of this call will be available to the client for each update unit.
+  Validation of the values is expected to return as quickly as possible and only be related to accepting or rejecting the set request.
+  It should not take into account how the changed value may or may not affect ongoing system performance of the node or the greater system.
+- **List Parameters**
+  Provide a list of parameter names which are currently set.
+- **Describe Parameters**
+  Given a list of parameter names, return their datatype.
+
+This functionality will be exposed through a user API which will support both local API calls as well as invocations on remote nodes via a ROS Service API.
+
+### Parameter update validation
+
+The node will allow for the registration of a callback for custom parameter value validation.
+
+### Backwards compatibility Parameter Server like behavior
+
+There are use cases where the older behavior with parameter server was useful.
+Both persisting beyond the duration of a specific node is valuable as well as having parameters with no specific association to a node which would potentially own or validate the values.
+To this end we propose to write a simple node which emulates the policy of the ROS 1.0 parameter server: it runs in namespace `/` and simply accepts all changes requested.
+The parameters held by this parameter server node would persist for the lifetime of the parameter server node.
+Specific instances could be launched in different namespaces to support different parameter persistence models.
+
+### Search parameter behavior
+
+A pattern developed in ROS 1.0 was the `searchParam` mode where a parameter could be set in a namespace and the parameter query would walk up the namespace to search for the parameter.
+A similar behavior can be implemented by allowing the search parameter implementation to walk across the different nodes in hierarchical order.
+
+
+### Parameter API
+
+The client libraries will provide an API for interfacing with the Core Parameter API for both local and remote nodes including return codes.
```

Comment:
This sentence seems to say almost the same as the last paragraph of "Required functionality" (https://github.com/ros2/design/pull/58/files#diff-4075c1e78328ec132523517becd96e1bR141).